packages/outbox-core/lib/outbox.ts

```diff
@@ -0,0 +1,127 @@
+ */
+export class OutboxProcessor<SupportedEvents extends CommonEventDefinition[]> {
+  constructor(
+    private readonly outboxStorage: OutboxStorage<SupportedEvents>,
+    private readonly outboxAccumulator: OutboxAccumulator<SupportedEvents>,
+    private readonly eventEmitter: DomainEventEmitter<SupportedEvents>,
+    private readonly maxRetryCount: number,
+    private readonly emitBatchSize: number,
+  ) {}
+
+  public async processOutboxEntries(context: JobExecutionContext) {
+    const entries = await this.outboxStorage.getEntries(this.maxRetryCount)
+
+    await PromisePool.for(entries)
+      .withConcurrency(this.emitBatchSize)
+      .process(async (entry) => {
+        try {
+          await this.eventEmitter.emit(entry.event, entry.data, entry.precedingMessageMetadata)
+          await this.outboxAccumulator.add(entry)
+        } catch (e) {
+          context.logger.error({ error: e }, 'Failed to process outbox entry.')
+
+          await this.outboxAccumulator.addFailure(entry)
+        }
+      })
+
+    await this.outboxStorage.flush(this.outboxAccumulator)
+    await this.outboxAccumulator.clear()
+  }
+}
+
+/**
+ * Periodic job that processes outbox entries every "intervalInMs". If processing takes longer than defined interval, another subsequent job WILL NOT be started.
+ *
+ * Each entry is ACKed, then event is published, and then entry is marked as SUCCESS. If processing fails, entry is marked as FAILED and will be retried.
+ *
+ * Max retry count is defined by the user.
+ */
+export class OutboxPeriodicJob<
+  SupportedEvents extends CommonEventDefinition[],
+> extends AbstractPeriodicJob {
+  private readonly outboxProcessor: OutboxProcessor<SupportedEvents>
+
+  constructor(
+    outboxStorage: OutboxStorage<SupportedEvents>,
+    outboxAccumulator: OutboxAccumulator<SupportedEvents>,
+    eventEmitter: DomainEventEmitter<SupportedEvents>,
+    dependencies: PeriodicJobDependencies,
+    maxRetryCount: number,
+    emitBatchSize: number,
+    intervalInMs: number,
+  ) {
+    super(
+      {
+        jobId: 'OutboxJob',
+        schedule: {
+          intervalInMs: intervalInMs,
+        },
+        singleConsumerMode: {
+          enabled: true,
+        },
+      },
+      {
+        redis: dependencies.redis,
+        logger: dependencies.logger,
+        transactionObservabilityManager: dependencies.transactionObservabilityManager,
+        errorReporter: dependencies.errorReporter,
+        scheduler: dependencies.scheduler,
+      },
+    )
+
+    this.outboxProcessor = new OutboxProcessor<SupportedEvents>(
+      outboxStorage,
+      outboxAccumulator,
+      eventEmitter,
+      maxRetryCount,
+      emitBatchSize,
+    )
+  }
+
+  protected async processInternal(context: JobExecutionContext): Promise<void> {
+    await this.outboxProcessor.processOutboxEntries(context)
+  }
+}
+
+export class OutboxEventEmitter<SupportedEvents extends CommonEventDefinition[]> {
```

Comment:
sounds a bit like a misnomer, as we are not emitting any events here, we are only scheduling them for future emission